### PR TITLE
Added Shell parameters

### LINKED ISSUE #6848: [Enhancement] Shell Navigation with Models
## Summary
Use a Model Class instead of using Query Property to pass data to a page.

## API Changes

In order to facilitate the new _Navigation_ api, an _object_ is added to the parameters of the _GoToAsync_ call.

    var model = new MyModel();
    await Shell.Current.GoToAsync("PAGE_URI", model); // call GoToAsync() using a Model Class

## Intended Use Case
This new feature could be useful for passing many complex data between pages, without having to pass all the data as strings.

### PR DESCRIPTION
**Pull request by pictos**
### Description of Change ###

This PR adds support for users passes complex data in Shell navigation with the new object `ShellParameter`.

### Issues Resolved ### 

- fix #6848

### API Changes ###

Added:
 - Shell.GoToAsync(ShellNavigationState state, IDictionary<string,object> parameters)
 - Shell.GoToAsync(ShellNavigationState state, bool animated, IDictionary<string,object> parameters)

 
### Platforms Affected ### 

- Core/XAML (all platforms)

### Behavioral/Visual Changes ###

None

### Before/After Screenshots ### 

Not applicable

### Testing Procedure ###

When you navigate with Shell you can pass `ShellParameter` in the method, just like:

`await Shell.Current.GoToAsync("//myAwesomeUri", new Dictionary { {"My key", someObj} });`

### PR Checklist ###

- [ ] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

### CODE REVIEW

File: Xamarin.Forms.Core/Shell/ShellParameter.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+			if (input == null || input.Count == 0)
+				return;
+
+			foreach (var item in input)
+				Add(item.Key, item.Value);
+		}
+	}
+
+
+	public static class ShellParameterExtensions
```
Comment (pictos) [inline Xamarin.Forms.Core/Shell/ShellParameter.cs:29]: I didn't find a good place to add this class. Can I create an extension folder?

File: Xamarin.Forms.Core/Shell/Shell.cs [hunk 2]
```diff
@@ -484,24 +484,25 @@ internal Task CompleteDeferredNavigating(ShellNavigatingEventArgs deferredArgs)
 			return GoToAsync(deferredArgs.Target, deferredArgs.Animate, false, deferredArgs);
 		}
 
-		public Task GoToAsync(ShellNavigationState state)
+		public Task GoToAsync(ShellNavigationState state, ShellParameter parameters = null)
```
Comment (PureWeen) [inline Xamarin.Forms.Core/Shell/Shell.cs:487]: Just adding optional parameters is a breaking change you have to add each overload manually

https://haacked.com/archive/2010/08/10/versioning-issues-with-optional-arguments.aspx/

**Review by PureWeen — CHANGES_REQUESTED** [review]

**Comment by StephaneDelcroix** [comment]
this removes the ability of serializing navigation Uris to strings (for, e.g., deep linking). I'd be more confortable if this was a Dictionary<string,string>

**Comment by StephaneDelcroix** [comment]
> * This obsoletes the `IQueryAttributable` interface that no one uses because it's not documented anywhere

that's a lame excuse for obsoleting :)

**Comment by PureWeen** [comment]
Apologies on this one @pictos 

I gave you premature guidance on this one and as it currently stands we can't really take in an API change like this pre .NET MAUI

**Comment by PureWeen** [comment]
@pictos can you fix the conflicts?

**Comment by pictos** [comment]
Yes, is it ok to solve it by tomorrow?

File: Xamarin.Forms.Core/Shell/Shell.cs [hunk 3]
```diff
@@ -495,6 +495,26 @@ public Task GoToAsync(ShellNavigationState state, bool animate)
 			return _navigationManager.GoToAsync(state, animate, false);
 		}
 
+		internal Task GoToAsync(ShellNavigationState state, ShellRouteParameters parameters)
```
Comment (pictos) [inline Xamarin.Forms.Core/Shell/Shell.cs:498]: @PureWeen not sure if we still need those methods here...

**Comment by PureWeen** [comment]
Re-closing this one again unfortunately :-/ 

I've been chatting with @pictos  a bit of how to bring this behavior to XCT

**Comment by angelru** [comment]
will it be available in XCT?

**Pull request closed without merging**